changefeedccl: support attributes in pubsub sink

jayshrivastava opened the pull request:
### changefeedccl: refactor EmitRow to take row metadata struct

This change updates the `EmitRow` method on the `EventSink` interface
to take a metadata struct instead of an mvcc timestamp and update timestamp.
This struct contains the mvcc and update timestamps such as before but
now also includes the table name corresponding to the row.

Release note: None
Epic: None

---

### support attributes in pubsub sink

This change adds support for pubsub attributes. Attributes
are essentially a `map[string]string` that is passed long with
each message.

These attributes can be specified via a new field in the `pubsub_sink_config` like so:
```
CREATE CHANGEFEED FOR ... WITH pubsub_sink_config='{"attributes": {"table": "TABLE_NAME", "foo": "bar"}}'
```

Keys and values can be arbitrary strings. Certain values are special, for
example `TABLE_NAME`. Specifying that value will result it in being
replaced with the table name corresponding to the row.

Version gates are not required. The `attributes` field should be ignored by nodes running
a version prior to this change. Those nodes will simply not emit attributes. A node running
this change will emit attributes and be able to resolve the `TABLE_NAME` given the previous
commit.

Release note (enterprise change): Pubsub attributes are now supported by changefeeds.
To add attributes to messages in a pubsub changefeed, add the "attributes" key to the existing
`pubsub_sink_config` changefeed option. For example:
```
CREATE CHANGEFEED FOR ... WITH pubsub_sink_config='{"attributes": {"table": "TABLE_NAME", "foo": "bar"}}'
```
The map of attributes will attached to each pubsub message sent from CockroachDB.
Certain attributes are special. Passing the "TABLE_NAME" value will result in the value
being replaced by the name of the table which the row was written to. Table names
will be lowercase.

jayshrivastava committed:
changefeedccl: support attributes in pubsub sink

This change adds support for pubsub attributes. Attributes
are essentially a `map[string]string` that is passed long with
each message.

These attributes can be specified via a new field in the `pubsub_sink_config` like so:
```
CREATE CHANGEFEED FOR ... WITH pubsub_sink_config='{"attributes": {"table": "TABLE_NAME", "foo": "bar"}}'
```

Keys and values can be arbitrary strings. Certain values are special, for
example `TABLE_NAME`. Specifying that value will result it in being
replaced with the table name corresponding to the row.

Version gates are not required. The `attributes` field should be ignored by nodes running
a version prior to this change. Those nodes will simply not emit attributes. A node running
this change will emit attributes and be able to resolve the `TABLE_NAME` given the previous
commit.

Release note (enterprise change): Pubsub attributes are now supported by changefeeds.
To add attributes to messages in a pubsub changefeed, add the "attributes" key to the existing
`pubsub_sink_config` changefeed option. For example:
```
CREATE CHANGEFEED FOR ... WITH pubsub_sink_config='{"attributes": {"table": "TABLE_NAME", "foo": "bar"}}'
```
The map of attributes will attached to each pubsub message sent from CockroachDB.
Certain attributes are special. Passing the "TABLE_NAME" value will result in the value
being replaced by the name of the table which the row was written to. Table names
will be lowercase.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111958)
<!-- Reviewable:end -->

jayshrivastava committed:
changefeedccl: refactor `EmitRow` to take row metadata struct

This change updates the `EmitRow` method on the `EventSink` interface
to take a metadata struct instead of an mvcc timestamp and update timestamp.
This struct contains the mvcc and update timestamps such as before but
now also includes the table name corresponding to the row.

Release note: None
Epic: None

jayshrivastava commented:
Closing in favor of https://github.com/cockroachdb/cockroach/pull/116089
